### ISSUE TITLE
Use ElasticSearch index aliases to avoid search downtime

### ISSUE THREAD

**Issue by stof** [mdn/yari#3098]
An option to avoid downtime while doing a full recreation of the index would be to use an [index alias](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-add-alias.html) for the *reading* usage of elasticsearch in the search endpoint.
The script would create a new index (probably with the date of the build in the index name), index all new content in it. Then it would recreate the alias to point to the new index. And then it can delete the old index which is not used anymore.

This way, there is no downtime. While the new index is populated, the search endpoint are still working, as they are still seeing the old data (as the alias still points to the old index at that point).

**Comment by peterbe** [comment]
I like this idea. But before we jump into this we should assert that it's possible in a hosted Elasticsearch. 

Also, @stof at the moment, there's been quite a lot of interesting new changes to the way we index so it's quite nice that any new mappings gets in within 24h. But a very low-tech solution we've considered is to make the decision whether to use `--update` or not use `--update` can be decided by if it's a Sunday. 
We run the indexing cron job on a daily basis. The `--update` flag just "cakes on" without regard for what might need to be deleted since it's no longer built. But the MDN traffic on a weekend is minuscule compared weekdays. 
That solution might require some engineering which might not be much less than employing your idea.

**Comment by stof** [comment]
> I like this idea. But before we jump into this we should assert that it's possible in a hosted Elasticsearch.

to be checked with your own hosting. As I don't know which one you use, I cannot answer for you.
If the hosting used is the one from Elastic, the answer is yes though.

**Comment by peterbe** [comment]
> If the hosting used is the one from Elastic, the answer is yes though.

That's what we use! Excellent. Thanks.

**Comment by peterbe** [comment]
I played with this today. And failed. 
First of all, even if you override the index name in the `streaming_bulk` indexing, it will still index the documents under the name of the `Document.Index.name` attribute. Weird. 
I essentially indexed everything under an index called `mdn_docs_20210326` and then I created an alias to that index called `mdn_docs`. Now I should, in theory, not have to change anything on the search side. But to my surprise, I kept getting empty search results all the time. But I could see that there were over 11k documents in the index. It was only when I removed the term filtering on the `locale` that it started to yield results. How bizarre. By giving up and going back to `main` and re-indexing as per usual, now searches worked perfectly again. 

I think I need to read up on how to create aliases because there might be more to it.
